example/vectorAdd/src/main.cpp

```diff
@@ -0,0 +1,201 @@
+/**
+ * \file
+ * Copyright 2014-2015 Benjamin Worpitz
+ *
+ * This file is part of alpaka.
+ *
+ * alpaka is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * alpaka is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with alpaka.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <alpaka/alpaka.hpp>                        // alpaka::exec::create
+
+#include <iostream>                                 // std::cout
+#include <typeinfo>                                 // typeid
+
+//#############################################################################
+//! A vector addition kernel.
+//! \tparam TAcc The accelerator environment to be executed on.
+//#############################################################################
+class VectorAddKernel
+{
+public:
+    //-----------------------------------------------------------------------------
+    //! The kernel entry point.
+    //!
+    //! \tparam TElem The matrix element type.
+    //! \param acc The accelerator to be executed on.
+    //! \param A The first source vector.
+    //! \param B The second source vector.
+    //! \param C The destination vector.
+    //! \param numElements The number of elements.
+    //-----------------------------------------------------------------------------
+    ALPAKA_NO_HOST_ACC_WARNING
+    template<
+        typename TAcc,
+        typename TElem,
+        typename TSize>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const & acc,
+        TElem const * const A,
+        TElem const * const B,
+        TElem * const C,
+        TSize const & numElements) const
+    -> void
+    {
+        static_assert(
+            alpaka::dim::Dim<TAcc>::value == 1,
+            "The VectorAddKernel expects 1-dimensional indices!");
+
+        auto const gridThreadIdx(alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]);
+        auto const threadElemExtent(alpaka::workdiv::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
+        auto const threadFirstElemIdx(gridThreadIdx * threadElemExtent);
+
+        if(threadFirstElemIdx < numElements)
+        {
+            // Calculate the number of elements to compute in this thread.
+            // The result is uniform for all but the last thread.
+            auto const elems(threadElemExtent + alpaka::math::min(acc, static_cast<unsigned long long>(0u), static_cast<unsigned long long>(numElements-(threadFirstElemIdx+threadElemExtent))));
```

Comment:
independent of the point that we should move this in a separate pull request:

Might be we should create a alpaka trait/function for this line.